### PR TITLE
Update export types

### PR DESCRIPTION
**Pull request by dbritto-dev**
## Related Issues or Discussions

Fixes #2163

## Summary

The types for should be under each type of import. For instance "import", "module", "require" and "default"

## Check List

- [x] `yarn run prettier` for formatting code and docs

### CODE REVIEW

**Comment by dbritto-dev** [comment]
@dai-shi here you go

**Comment by dai-shi** [comment]
> ## Related Issues or Discussions
> Fixes #2164

Do you mean #2163 ?

File: package.json [hunk 1]
```diff
@@ -21,12 +21,13 @@
   "exports": {
     "./package.json": "./package.json",
     ".": {
+      "types": "./esm/index.d.ts",
```
Comment (dai-shi) [inline package.json:24]: Do you need this? Also, I just noticed that we didn't change other than ".".

Comment (dbritto-dev) [reply package.json:24]: For some reason esm types throw some errors, that set up the `index.d.ts` as default for everything but internally we use the ones that we had under types on each config

**Comment by dbritto-dev** [comment]
> > ## Related Issues or Discussions
> 
> > Fixes #2164
> 
> 
> 
> Do you mean #2163 ?

Yeah, that one

**Comment by dbritto-dev** [comment]
@dai-shi all yours

File: package.json [hunk 2]
```diff
@@ -26,85 +26,27 @@
...
-      "import": {
-        "types": "./esm/react/shallow.d.mts",
-        "default": "./esm/react/shallow.mjs"
-      },
-      "module": "./esm/react/shallow.js",
-      "default": "./react/shallow.js"
-    },
-    "./traditional": {
-      "types": "./traditional.d.ts",
+    "./*": {
```
Comment (dai-shi) [inline package.json:37]: This requires newer version of Node, and we will be doing this change in v5.

File: package.json [hunk 3]
```diff
@@ -26,85 +26,27 @@
         "default": "./esm/index.mjs"
       },
       "module": {
-        "types": "./index.d.ts",
+        "types": "./esm/index.d.ts",
```
Comment (dai-shi) [inline package.json:29]: Is this change the key to fix the issue? I mean the file content must be the same, so the file path was the root cause, right? Why wasn't it an issue before #2157?

It feels like reverting #2157 should be appropriate, if we **don't know** why it caused such an issue.

**Review by dai-shi — CHANGES_REQUESTED** [review]

File: package.json (hunk 3 above)
Comment (dbritto-dev) [reply package.json:29]: @dai-shi btw, I updated the exports to match jotai's and valtio's exports

Comment (dbritto-dev) [reply package.json:29]: @dai-shi but we can revert the changes

Comment (dai-shi) [reply package.json:29]: Yeah > https://github.com/pmndrs/zustand/pull/2170#discussion_r1379421218

File: package.json (hunk 2 above)
Comment (dai-shi) [reply package.json:37]: @dbritto-dev We are waiting for the revert.

Comment (dbritto-dev) [reply package.json]: @dai-shi done

Comment (dai-shi) [reply package.json]: Ooops, sorry. I don't mean to revert #2157 fully.
Can you try what you think should work, and just avoid using `"./*"`.

Comment (dbritto-dev) [reply package.json]: In that we case we should apply the same change for each inner package until we can use the `./*`

Comment (dai-shi) [reply package.json]: Yes, that what I meant.

Comment (dbritto-dev) [reply package.json]: Sure, let me do it

**Comment by dai-shi** [comment]
btw, @dbritto-dev do you understand why #2157 caused such an issue? It seems like TS picks different files for index and middleware, but I'm not sure how that could happen. It sounds like TS ignore "types" field, doesn't it?

**Comment by dbritto-dev** [comment]
> btw, @dbritto-dev do you understand why #2157 caused such an issue? It seems like TS picks different files for index and middleware, but I'm not sure how that could happen. It sounds like TS ignore "types" field, doesn't it?

@dai-shi I understand the issue in the case of inner packages when you use "types" outside of "import", "module", "require" or "default" ts use by default that types ignoring where you are importing the packages when we added the new changes for the main package we fixed that but the inner packages still pointing to "d.ts" even if you are importing them from "esm" that's the conflict

**Comment by dbritto-dev** [comment]
@dai-shi all yours

**Comment by dai-shi** [comment]
> in the case of inner packages when you use "types" outside of "import", "module", "require" or "default" ts use by default that types ignoring where you are importing the packages;

So, the "types" line right under "exports" in v4.4.4 package.json was simply ignored? That explains it.

**Comment by dbritto-dev** [comment]
@dai-shi would you mind taking care of these changes? I tested locally and works well

**Comment by dai-shi** [comment]
Yes, it looks good.
I hope to get one or two more people to try it. https://github.com/pmndrs/zustand/discussions/2163#discussioncomment-7457555

**Review by dai-shi — APPROVED** [review]
The changes look good.

**Comment by LarsArtmann** [comment]
Thank you for fixing my build :) 👍

**Comment by sanduluca** [comment]
Was this released as I still have the issue with immer and `"zustand": "^5.0.3"`

image

**Comment by dbritto-dev** [comment]
> Was this released as I still have the issue with immer and `"zustand": "^5.0.3"`
> 
> image

@sanduluca that doesn't look right to me, `addSound` method should be implemented keeping in mind `immer`

**Comment by sanduluca** [comment]
In the image i was showing the that the immer is generating the error (before refactoring). 

But the point is still that it shows type errors with immer even with the example from docs.

image

<details><summary>Code and error message</summary>
<p>

```ts
type State = {
  count: number;
};

type Actions = {
  increment: (qty: number) => void;
  decrement: (qty: number) => void;
};

export const useCountStore = create<State & Actions>()(
  immer((set) => ({
    count: 0,
    increment: (qty: number) =>
      set((state) => {
        state.count += qty;
      }),
    decrement: (qty: number) =>
      set((state) => {
        state.count -= qty;
      }),
  }))
);
```
```bash
Argument of type 'StateCreator<State & Actions, [], [["zustand/immer", never]]>' is not assignable to parameter of type 'StateCreator<State & Actions, [], [never, unknown][]>'.
  Type 'StateCreator<State & Actions, [], [["zustand/immer", never]]>' is not assignable to type '{ $$storeMutators?: [never, unknown][] | undefined; }'.
    Types of property '$$storeMutators' are incompatible.
      Type '[["zustand/immer", never]] | undefined' is not assignable to type '[never, unknown][] | undefined'.
        Type '[["zustand/immer", never]]' is not assignable to type '[never, unknown][]'.
          Type '["zustand/immer", never]' is not assignable to type '[never, unknown]'.
            Type at position 0 in source is not compatible with type at position 0 in target.
              Type '"zustand/immer"' is not assignable to type 'never'.
```

image

</p>
</details>

**Comment by dbritto-dev** [comment]
> In the image i was showing the that the immer is generating the error (before refactoring). 
> 
> 
> 
> But the point is still that it shows type errors with immer even with the example from docs.
> 
> 
> 
> image
> 
> 
> 
> <details><summary>Code, error message, package.json</summary>
> 
> <p>
> 
> 
> 
> ```ts
> 
> type State = {
> 
>   count: number;
> 
> };
> 
> 
> 
> type Actions = {
> 
>   increment: (qty: number) => void;
> 
>   decrement: (qty: number) => void;
> 
> };
> 
> 
> 
> export const useCountStore = create<State & Actions>()(
> 
>   immer((set) => ({
> 
>     count: 0,
> 
>     increment: (qty: number) =>
> 
>       set((state) => {
> 
>         state.count += qty;
> 
>       }),
> 
>     decrement: (qty: number) =>
> 
>       set((state) => {
> 
>         state.count -= qty;
> 
>       }),
> 
>   }))
> 
> );
> 
> ```
> 
> ```bash
> 
> Argument of type 'StateCreator<State & Actions, [], [["zustand/immer", never]]>' is not assignable to parameter of type 'StateCreator<State & Actions, [], [never, unknown][]>'.
> 
>   Type 'StateCreator<State & Actions, [], [["zustand/immer", never]]>' is not assignable to type '{ $$storeMutators?: [never, unknown][] | undefined; }'.
> 
>     Types of property '$$storeMutators' are incompatible.
> 
>       Type '[["zustand/immer", never]] | undefined' is not assignable to type '[never, unknown][] | undefined'.
> 
>         Type '[["zustand/immer", never]]' is not assignable to type '[never, unknown][]'.
> 
>           Type '["zustand/immer", never]' is not assignable to type '[never, unknown]'.
> 
>             Type at position 0 in source is not compatible with type at position 0 in target.
> 
>               Type '"zustand/immer"' is not assignable to type 'never'.
> 
> ```
> 
> 
> 
> image
> 
> 
> 
> </p>
> 
> </details> 

@sanduluca would you mind opening a discussion?

**Comment by Ray5055** [comment]
Hi! Is there any updates on this matter? I'm facing similar issues however mine is with persist rather than immer

```
Argument of type 'StateCreator<AssetState, [], [["zustand/persist", AssetState]]>' is not assignable to parameter of type 'StateCreator<AssetState, [], [never, unknown][]>'.
  Type 'StateCreator<AssetState, [], [["zustand/persist", AssetState]]>' is not assignable to type '{ $$storeMutators?: [never, unknown][] | undefined; }'.
    Types of property '$$storeMutators' are incompatible.
      Type '[["zustand/persist", AssetState]] | undefined' is not assignable to type '[never, unknown][] | undefined'.
        Type '[["zustand/persist", AssetState]]' is not assignable to type '[never, unknown][]'.
          Type '["zustand/persist", AssetState]' is not assignable to type '[never, unknown]'.
            Type at position 0 in source is not compatible with type at position 0 in target.
              Type '"zustand/persist"' is not assignable to type 'never'.ts(2345)
```

**Pull request merged**